packages/vats/src/orchestration.js

```diff
@@ -0,0 +1,255 @@
+ * >} PowerStore
+ */
+
+/**
+ * @template {keyof OrchestrationPowers} K
+ * @param {PowerStore} powers
+ * @param {K} name
+ */
+const getPower = (powers, name) => {
+  powers.has(name) || Fail`need powers.${bare(name)} for this method`;
+  return /** @type {OrchestrationPowers[K]} */ (powers.get(name));
+};
+
+export const ChainAccountI = M.interface('ChainAccount', {
+  getAddress: M.call().returns(M.string()),
+  getLocalAddress: M.callWhen().returns(M.string()),
+  executeTx: M.callWhen(M.arrayOf(M.record())).returns(M.any()),
+  executeEncodedTx: M.callWhen(M.string()).returns(M.any()),
+  deposit: M.callWhen(M.remotable('Payment'))
+    .optional(M.pattern())
+    .returns(AmountShape),
+  getPurse: M.callWhen(BrandShape).returns(M.remotable('VirtualPurse')),
+  close: M.callWhen().returns(M.promise()),
+});
+
+/** @param {import('@agoric/base-zone').Zone} zone */
+const prepareChainAccount = zone =>
+  zone.exoClass(
+    'ChainAccount',
+    ChainAccountI,
+    /**
+     * @param {Connection} connection
+     * @param {Port} port
+     * @param {ChainAddress} remoteAddress
+     */
+    (connection, port, remoteAddress) => ({
+      connection,
+      port,
+      remoteAddress,
+    }),
+    {
+      getAddress() {
+        return this.state.remoteAddress;
+      },
+      getLocalAddress() {
+        return E(this.state.connection).getLocalAddress();
+      },
+      /** @param {Bytes} packetBytes */
+      async executeEncodedTx(packetBytes) {
+        return E(this.state.connection).send(packetBytes);
+      },
+      /** @param {import('@agoric/vats/src/localchain').Proto3Jsonable[]} _msgs */
+      async executeTx(_msgs) {
+        // XXX encode to protobuf, and call this.executeEncodedTx(...)
+        Fail`executeTx not implemented yet`;
+      },
+      async deposit(_payment) {
+        // XXX deposit an ERTP payment to the remote account
+        Fail`deposit not implemented yet`;
+      },
+      async getPurse(_brand) {
+        Fail`getPurse not implemented yet`;
+      },
+      async close() {
+        // XXX retrieve all assets first?
+        // XXX do we also revoke the port?
+        return E(this.state.connection).close();
+      },
+    },
+  );
+
+/** @param {import('@agoric/zone').Zone} zone */
+const prepareConnectionHandler = zone =>
+  zone.exoClass(
+    'ConnectionHandler',
+    undefined, // XXX interface guard
+    /**
+     * @param {(
+     *   connection: Connection,
+     *   localAddr: string,
+     *   remoteAddr: string,
+     * ) => void} [onOpen]
+     *   optional cb handler for connection creation
+     * @param {(connection: Connection, reason: unknown) => void} [onClose]
+     *   optional cb handler for connection closing
+     * @param {(
+     *   connection: Connection,
+     *   bytes: Uint8Array,
+     * ) => PromiseVow<string>} [onReceive]
+     *   optional cb handler for connection recieving packet
+     */
+    (onOpen, onClose, onReceive) => ({ onOpen, onClose, onReceive }),
```

Comment:
storing functions in durable state? does that work? I'd be surprised.

Reply:
ty, addressed